### ISSUE TITLE
Can't unselect text or select specific text neither see the text cursor (caret) when using SearchBar

### ISSUE THREAD

**Issue by lucasoares** [flutter/flutter#131235]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

1. Create a SearchBar inside a SearchAnchor
2. Click on the search bar
3. Write something
4. Try to select some of the text
5. Close the search view keeping the text
6. Click on the search bar
7. Try to unselect the text (click in the text like you want to make the caret go to the end of the text)

If you implement a FocusNode for the SearchBar and make it request focus, the caret will not appear, only after pressing the arrow right key.

After closing the view, the text should not be selected too. Check the video for example.

Also, there is no way to configure the 'enter' keypress in the search view to close the view and submit the search.

### Expected results

The UX should be exactly like the search bar of common sites, like https://www.google.com/

### Actual results

Bad UX experience, can't select specific parts of the text, can't unselect the text without pressing the right arrow key (on the Web, for example)

### Code sample

<details><summary>Code sample</summary>

THIS IS THE EXAMPLE CODE

```dart
// Copyright 2014 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'package:flutter/material.dart';

/// Flutter code sample for [SearchBar].

void main() => runApp(const SearchBarApp());

class SearchBarApp extends StatefulWidget {
  const SearchBarApp({super.key});

  @override
  State<SearchBarApp> createState() => _SearchBarAppState();
}

class _SearchBarAppState extends State<SearchBarApp> {
  bool isDark = false;

  @override
  Widget build(BuildContext context) {
    final ThemeData themeData = ThemeData(useMaterial3: true, brightness: isDark ? Brightness.dark : Brightness.light);

    return MaterialApp(
      theme: themeData,
      home: Scaffold(
        appBar: AppBar(title: const Text('Search Bar Sample')),
        body: Padding(
          padding: const EdgeInsets.all(8.0),
          child: SearchAnchor(
            builder: (BuildContext context, SearchController controller) {
              return SearchBar(
                controller: controller,
                padding: const MaterialStatePropertyAll<EdgeInsets>(EdgeInsets.symmetric(horizontal: 16.0)),
                onTap: () { controller.openView(); },
                onChanged: (_) { controller.openView(); },
                leading: const Icon(Icons.search),
                trailing: <Widget>[
                  Tooltip(
                    message: 'Change brightness mode',
                    child: IconButton(
                      isSelected: isDark,
                      onPressed: () {
                        setState(() {
                          isDark = !isDark;
                        });
                      },
                      icon: const Icon(Icons.wb_sunny_outlined),
                      selectedIcon: const Icon(Icons.brightness_2_outlined),
                    ),
                  )
                ],
              );
            },
            suggestionsBuilder: (BuildContext context, SearchController controller) {
              return List<ListTile>.generate(5, (int index) {
                final String item = 'item $index';
                return ListTile(
                  title: Text(item),
                  onTap: () {
                    setState(() {
                      controller.closeView(item);
                    });
                  },
                );
              });
            }),
        ),
      ),
    );
  }
}

```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

After clicking the search bar, all text is selected:
![image](https://github.com/flutter/flutter/assets/10624972/a9002b0c-eb9b-4377-81a4-0c05f7d9fbbc)

> Clicking anywhere in the text remains selected, even clicking at the end of the text (like normal text inputs should behave).

Video demonstration:

https://github.com/flutter/flutter/assets/10624972/e9584789-0d73-4e91-a8c9-dbfbe3409369

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

> Using BETA channel because of another bug of the SearchBar: https://github.com/flutter/flutter/issues/126623

> Not using Linux toolchain

```console
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel beta, 3.13.0-0.2.pre, on Ubuntu 22.04.2 LTS 5.19.0-46-generic, locale en_US.UTF-8)
[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
[✓] Chrome - develop for the web
[✗] Linux toolchain - develop for Linux desktop
    ✗ ninja is required for Linux development.
      It is likely available from your distribution (e.g.: apt install ninja-build), or can be downloaded from
      https://github.com/ninja-build/ninja/releases
[✓] Android Studio (version 2022.2)
[✓] IntelliJ IDEA Community Edition (version 2023.1)
[✓] VS Code (version 1.81.0)
[✓] VS Code (version 1.81.0-insider)
[✓] Connected device (2 available)
[✓] Network resources

! Doctor found issues in 1 category.
```

</details>

**Comment by huycozy** [comment]
Thanks for your report. Reproduced the issue on the latest Flutter channels. But it only appears on Web platform whilst it works as expected on Android/MacOS platforms.

<details>
<summary>flutter doctor -v (stable and master)</summary>

```bash
[✓] Flutter (Channel stable, 3.10.6, on macOS 13.0.1 22A400 darwin-x64, locale en-VN)
    • Flutter version 3.10.6 on channel stable at /Users/huynq/Documents/GitHub/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f468f3366c (5 hours ago), 2023-07-12 15:19:05 -0700
    • Engine revision cdbeda788a
    • Dart version 3.0.6
    • DevTools version 2.23.1

[✓] Android toolchain - develop for Android devices (Android SDK version 32.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-33, build-tools 32.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E222b
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.80.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.68.0

[✓] Connected device (3 available)
    • RMX2001 (mobile) • EUYTFEUSQSRGDA6D • android-arm64  • Android 11 (API 30)
    • macOS (desktop)  • macos            • darwin-x64     • macOS 13.0.1 22A400 darwin-x64
    • Chrome (web)     • chrome           • web-javascript • Google Chrome 114.0.5735.198

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

```bash
[!] Flutter (Channel master, 3.13.0-9.0.pre.43, on macOS 13.4.1 22F770820d darwin-x64, locale en-VN)
    • Flutter version 3.13.0-9.0.pre.43 on channel master at /Users/huynq/Documents/GitHub/flutter_master
    ! Warning: `flutter` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 0733c96e7d (4 hours ago), 2023-07-24 21:55:24 -0400
    • Engine revision 9a0192d965
    • Dart version 3.1.0 (build 3.1.0-333.0.dev)
    • DevTools version 2.25.0
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 32.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-33, build-tools 32.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E222b
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.80.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.68.0

[✓] Connected device (3 available)
    • RMX2001 (mobile) • EUYTFEUSQSRGDA6D • android-arm64  • Android 11 (API 30)
    • macOS (desktop)  • macos            • darwin-x64     • macOS 13.4.1 22F770820d darwin-x64
    • Chrome (web)     • chrome           • web-javascript • Google Chrome 114.0.5735.198

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```
</details>

**Comment by lucasoares** [comment]
Following the UX issues, pressing 'enter' key on Web is not submitting the search.

**Comment by lucasoares** [comment]
@huycozy there are few UX issues on android too! You can't move the cursor to write in the middle of text you've already written

**Comment by huycozy** [comment]
Thanks for checking that. 

I check this issue again on Android, iOS and notice the a little different in result:

- On Android: I see the cursor can move if I enter text into the search bar. But the cursor doesn't move if I select one of the suggestion items in its builder. 
- On iOS: The cursor can not move even with entered text or selected suggestion item from its builder. If I move cursor with spacebar, it works as expected (please watch demo for details)

<details>
<summary>Demo video (Android, iOS)</summary>

| Android | iOS |
| --------------- | --------------- |
<video src="https://github.com/flutter/flutter/assets/104349824/792b13ed-be3d-42d4-9b2f-be4440488f77"/> | <video src="https://github.com/flutter/flutter/assets/104349824/5105819b-4805-47bd-936a-6e049766b3bb"/>

</details>

Updating labels since it also affects other platforms rather than only Web.

**Comment by huycozy** [comment]
>if you use searchbar without opening the search view, you will see that is impossible selecting the text.

I can't imagine how that case looks like, can you share a demo with specific steps?

> If there is no suggestions to show it would be nice to not having to open the view for a better UX experience. Currently is like impossible because the input text of the searchbar itself is pretty much useless without opening the view.

This looks like a new proposal, please file a new issue for this.

**Comment by lucasoares** [comment]
Remove both the following lines:

```
                onTap: () { controller.openView(); },
                onChanged: (_) { controller.openView(); },
```

You will be able to use the search bar without opening the search view. This is not a new proposal, it's something that is already possible. The problem is that if you do remove these lines, you will not be able to do things we already spoken (move caret, etc) and the search bar input will have a bad UX experience.

This should work. It's not always that you should open the view (or if it is, the search bar should never let people to not open the view after tap). For exemple if there isn't a suggestion to show, there is no reason to open the view.

**Comment by justinmc** [comment]
CC @Renzo-Olivares, you said you had seen this before.

**Comment by Renzo-Olivares** [comment]
Closing as a duplicate of #128332

**Comment by lucasoares** [comment]
@Renzo-Olivares  I don't think this is a duplicate of #128332

SearchAnchor and SearchView are different widgets.

**Comment by Renzo-Olivares** [comment]
@lucasoares Hello the underlying issue is in `SearchBar` widget so this issue is a duplicate.
